x/psiphon/psiphon.go

```diff
@@ -0,0 +1,332 @@
+	pConfig.DisableLocalHTTPProxy = true
+	pConfig.DisableLocalSocksProxy = true
+	pConfig.TargetApiProtocol = "ssh"
+
+	err = pConfig.Commit(false)
+	if err != nil {
+		return fmt.Errorf("failed to commit config: %w", err)
+	}
+
+	// Will receive a value when the tunnel has successfully connected.
+	connected := make(chan struct{}, 1)
+	// Will receive a value if an error occurs during the connection sequence.
+	errCh := make(chan error, 1)
+
+	// Set up NoticeWriter to receive events.
+	psi.SetNoticeWriter(psi.NewNoticeReceiver(
+		func(notice []byte) {
+			var event clientlib.NoticeEvent
+			err := json.Unmarshal(notice, &event)
+			if err != nil {
+				// This is unexpected and probably indicates something fatal has occurred.
+				// We'll interpret it as a connection error and abort.
+				err = fmt.Errorf("failed to unmarshal notice JSON: %w", err)
+				select {
+				case errCh <- err:
+				default:
+				}
+				return
+			}
+			switch event.Type {
+			case "EstablishTunnelTimeout":
+				select {
+				case errCh <- context.DeadlineExceeded:
+				default:
+				}
+			case "Tunnels":
+				count := event.Data["count"].(float64)
+				if count > 0 {
+					select {
+					case connected <- struct{}{}:
+					default:
+					}
+				}
+			}
+		}))
+
+	err = psi.OpenDataStore(&psi.Config{DataRootDirectory: pConfig.DataRootDirectory})
+	if err != nil {
+		return fmt.Errorf("failed to open data store: %w", err)
+	}
+	needsCleanup := true
+	defer func() {
+		if needsCleanup {
+			psi.CloseDataStore()
+		}
+	}()
+
+	controller, err := psi.NewController(pConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create Controller: %w", err)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		controllerDone := make(chan struct{})
+		d.controllerDone = controllerDone
+		controller.Run(ctx)
+		close(controllerDone)
+	}()
+
+	// Wait for an active tunnel or error
+	select {
+	case <-connected:
+		needsCleanup = false
+		d.cancel = cancel
+		d.controller = controller
+		return nil
+	case err := <-errCh:
+		cancel()
+		if !errors.Is(err, context.DeadlineExceeded) {
+			err = fmt.Errorf("failed to run Controller: %w", err)
+		}
+		return err
+	}
+}
+
+// Stop stops the Dialer background processes, releasing resources and allowing it to be reconfigured.
+func (d *Dialer) Stop() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.cancel == nil || d.controller == nil {
+		return errors.New("tried to stop dialer that is not running")
+	}
+	d.cancel()
+	// Wait for Controller to finish before cleaning up.
+	<-d.controllerDone
+	d.cancel = nil
+	d.controller = nil
+	psi.CloseDataStore()
+	return nil
+}
+
+// GetSingletonDialer returns the single Psiphon dialer instance.
+func GetSingletonDialer() *Dialer {
+	return &singletonDialer
+}
+
+// streamConn wraps a [net.Conn] to provide a [transport.StreamConn] interface.
+type streamConn struct {
+	net.Conn
+}
+
+var _ transport.StreamConn = (*streamConn)(nil)
+
+func (c streamConn) CloseWrite() error {
+	return nil
```

Comment:
Shall we call `c.Close` somewhere? For example:

```
func (c streamConn) CloseWrite() error {
  c.writeClosed = true
  if c.writeClosed && c.readClosed {
    return c.Close()
  }
  return nil
}
```

Reply:
That's not the right CloseWrite semantics. It's just an indication the write is done, but we can still read.

The user still needs to call c.Close()